src/registry.rs

```diff
@@ -0,0 +1,268 @@
+    where
+        I: IntoIterator<Item = (K, V)>,
+        K: AsRef<str>,
+        V: Into<String>;
+
+    /// List all registered URLs
+    fn list_urls(&self) -> Vec<String>;
+}
+
+/// The default [`ObjectStoreRegistry`]
+pub struct DefaultObjectStoreRegistry {
+    /// A map from scheme to object store that serve list / read operations for the store
+    object_stores: RwLock<HashMap<String, Arc<dyn ObjectStore>>>,
+}
+
+impl std::fmt::Debug for DefaultObjectStoreRegistry {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let stores = self.object_stores.read().unwrap();
+        f.debug_struct("DefaultObjectStoreRegistry")
+            .field("schemes", &stores.keys().cloned().collect::<Vec<_>>())
+            .finish()
+    }
+}
+
+impl Default for DefaultObjectStoreRegistry {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+impl DefaultObjectStoreRegistry {
+    /// Create a new [`DefaultObjectStoreRegistry`] with no registered stores
+    pub fn new() -> Self {
+        let object_stores = RwLock::new(HashMap::new());
+        Self { object_stores }
+    }
+}
+
+///
+/// Stores are registered based on the scheme, host and port of the provided URL.
+///
+/// For example:
+///
+/// - `file:///my_path` will return the default LocalFS store
+/// - `s3://bucket/path` will return a store registered with `s3://bucket` if any
+/// - `hdfs://host:port/path` will return a store registered with `hdfs://host:port` if any
+impl ObjectStoreRegistry for DefaultObjectStoreRegistry {
+    fn register_store(
+        &self,
+        url: &Url,
+        store: Arc<dyn ObjectStore>,
+    ) -> Option<Arc<dyn ObjectStore>> {
+        let s = get_url_key(url);
+        let mut stores = self.object_stores.write().unwrap();
+        stores.insert(s, store)
+    }
+
+    fn get_store(&self, url: &Url) -> Option<Arc<dyn ObjectStore>> {
+        self.get_store_with_opts(url, std::iter::empty::<(&str, &str)>())
+    }
+
+    fn get_store_with_opts<I, K, V>(&self, url: &Url, opts: I) -> Option<Arc<dyn ObjectStore>>
+    where
+        I: IntoIterator<Item = (K, V)>,
+        K: AsRef<str>,
+        V: Into<String>,
+    {
+        let s = get_url_key(url);
+        let mut stores = self.object_stores.write().unwrap();
+
+        if let Some(store) = stores.get(&s) {
+            return Some(Arc::clone(store));
+        }
+
+        match parse_url_opts(url, opts) {
+            Ok((store, _)) => {
+                let store = Arc::from(store);
+                stores.insert(s, Arc::clone(&store));
+                Some(store)
+            }
+            Err(_) => None,
+        }
+    }
+
+    fn list_urls(&self) -> Vec<String> {
+        let stores = self.object_stores.read().unwrap();
+        stores.keys().cloned().collect()
+    }
+}
+
+/// Get the key of a url for object store registration.
+/// The credential info will be removed
+fn get_url_key(url: &Url) -> String {
```

Comment:
I recommend making this a public URL so that downstream consumers can utilize it for consistency as well

Reply:
It seems like an implementation detail of the `DefaultObjectStoreRegistry` to me 🤔  so I am not sure about making it public

If it needs to be public, perhaps it could a method on `DefaultObjectStoreRegistry` like:

```rust
impl DefaultObjectStoreRegistry {
  pub fn url_key(url: &Url) -> String { ... }
}
```

Reply:
Yea, IMO it should be private since it's just used for the default registry. Other registries might want to cache differently.